docs/tutorials/go-built-in.md

```diff
@@ -103,579 +103,517 @@ func NewKVStoreApplication() *KVStoreApplication {
-etc.) by Tendermint Core.
+When Tendermint Core receives a new transaction, Tendermint asks the application
+if the transaction is acceptable. In our new application, let's implement some
+basic validation for the transactions it will receive.
 
-Valid transactions will eventually be committed given they are not too big and
-have enough gas. To learn more about gas, check out ["the
-specification"](https://docs.tendermint.com/master/spec/abci/apps.html#gas).
+For our KV store application, a transaction is a string with the form `key=value`,
+indicating a key and value to write to the store.
 
-For the underlying key-value store we'll use
-[badger](https://github.com/dgraph-io/badger), which is an embeddable,
-persistent and fast key-value (KV) database.
+Add the following helper method to `app.go`:
 
 ```go
-import "github.com/dgraph-io/badger"
+func (app *KVStoreApplication) validateTx(tx []byte) uint32 {
+	parts := bytes.SplitN(tx, []byte("="), 2)
 
-type KVStoreApplication struct {
- db           *badger.DB
- currentBatch *badger.Txn
+	// check that the transaction is not malformed
+	if len(parts) != 2 {
+		return 1
+	}
+	return 0
 }
+```
 
-func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
- return &KVStoreApplication{
-  db: db,
- }
+And call it from within your `CheckTx` method:
+
+```go
+func (app *KVStoreApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
+	code := app.validateTx(req.Tx)
+	return abcitypes.ResponseCheckTx{Code: code}
 }
 ```
 
+Any response with a non-zero code will be considered invalid by Tendermint.
+Our `CheckTx` logic returns `0` to Tendermint when a transaction passes
+its validation checks. The specific value of the code is meaningless to Tendermint.
+Non-zero codes are logged by Tendermint so applications can provide more specific
+information on why the transaction was rejected.
+
+Note that `CheckTx` _does not execute_ the transaction, it only verifies that that the
+transaction _could_ be executed. We do not know yet if the rest of the network has
+agreed to accept this transaction into a block. 
+
+Finally, make sure to add the `bytes` package to the your import stanza
+at the top of `app.go`:
+
+```go
+import (
+ "bytes"
+
+ abcitypes "github.com/tendermint/tendermint/abci/types"
+)
+```
+
+
+While this `CheckTx` is simple and only validates that the transaction is well-formed,
+it is very common for `CheckTx` to make more complex use of the state of an application.
+
 ### 1.3.2 BeginBlock -> DeliverTx -> EndBlock -> Commit
 
-When Tendermint Core has decided on the block, it's transfered to the
-application in 3 parts: `BeginBlock`, one `DeliverTx` per transaction and
-`EndBlock` in the end. DeliverTx are being transfered asynchronously, but the
-responses are expected to come in order.
+When the Tendermint consensus engine has decided on the block, the block is transferred to the
+application over three ABCI method calls: `BeginBlock`, `DeliverTx`, and `EndBlock`.
```

Comment:
Probably something for a future iteration, but it occurs to me this is probably where we should be checking for duplicate keys. Otherwise, we accept a Tx only to fail when the block is committed, and the chain stalls.

Reply:
Not totally sure I understand the issue being described here?

Reply:
Actually I think this is no longer an issue. I was concerned because we were previously failing for txn that updated an existing key. Now that check is gone, it's probably not important to worry about here.

The corner case is smaller now: A database tx may get conflicts if we try to include multiple txn for the same key _in a single block_, but for demonstration that's not super critical.